src/rnnlm/sample_a_word.cc

```diff
@@ -0,0 +1,432 @@
+      HistType h(first, last);
+      int32 word_new = history.back();
+      HistType::const_iterator last_new = history.end() - 1;
+      HistType h_new(history.begin(), last_new);
+      prob = pow(10, GetBackoffWeight(order, word_new, h_new)) *
+             pow(10, GetProb(order, word, h));
+      (*pdf).push_back(prob);
+    }
+  }
+}
+
+// Get history weights
+void NgramModel::ComputeHistoriesWeights() {
+  for (auto it = histories_.begin(); it != histories_.end(); ++it) {
+    HistType history(*(it));
+    assert(history.size() <= ngram_order_);
+    for (int32 i = 0; i < history.size() + 1; i++) {
+      HistType h_tmp = history;
+      float prob = 1.0 / histories_.size();
+      while (h_tmp.size() > (history.size() - i)) {
+        HistType::iterator last = h_tmp.end() - 1;
+        HistType h(h_tmp.begin(), last);
+        int32 word = h_tmp.back();
+        prob *= pow(10, GetBackoffWeight(h_tmp.size(), word, h));
+        h_tmp = h;
+      }
+      HistType::iterator begin = history.begin() + i;
+      HistType h(begin, history.end());
+      hists_weights_[h] += prob;
+    }
+  } 
+  std::cout << "Size of hists_weights_ is: " << hists_weights_.size() << std::endl;
+}
+
+// Get weighted pdf
+void NgramModel::ComputeWeightedPdf(std::vector<float>* pdf_w) {
+  float prob = 0;
+  (*pdf_w).resize(vocab_size_); // if do not do this, (*pdf_w)[word] += prob will get seg fault
+  for (int32 i = 0; i < vocab_size_; i++) {
+    for (auto it = hists_weights_.begin(); it != hists_weights_.end(); ++it) {
+      HistType h(it->first);
+      int32 order = h.size();
+      auto it_hist = probs_[order].find(h);
+      if (it_hist != probs_[order].end()) {
+        int32 word = i;
+        auto it_word = probs_[order][h].find(word);
+        if (it_word != probs_[order][h].end()) {
+          if (order > 0) {
+            HistType::iterator last = h.end() - 1;
+            HistType::iterator first = h.begin() + 1;
+            HistType h1(h.begin(), last);
+            HistType h2(first, h.end());
+            prob = (it->second) * (pow(10, probs_[order][h][word].first) - 
+                    pow(10, GetBackoffWeight(order, h.back(), h1))
+                    * pow(10, GetProb(order, word, h2)));
+            (*pdf_w)[word] += prob;
+          } 
+          else {
+            prob = (it->second) * pow(10, probs_[order][h][word].first);
+            (*pdf_w)[word] += prob;
+          }
+        }
+      }
+    } // end reading history
+  } // end reading words
+}
+
+// sample a word that follows a pdf
+int32 NgramModel::SampleWord(const std::vector<float>& pdf) {
+  // generate a cdf from the given pdf 
+  std::vector<std::pair<float, float> > cdf;
+  float upper = 0;
+  float lower = 0;
+  std::pair<float, float> probs;
+  for (int32 i = 0; i < pdf.size(); i++) {
+    upper += pdf[i];
+    lower = upper - pdf[i];
+    probs = std::make_pair(lower, upper);
+    cdf.push_back(probs);
+  }
+  float u = 1.0 * rand()/RAND_MAX;
```

Comment:
Did you try searching for the previous emails?
If you still can't find it, I will show you how to find a function in person later today.